replication/logic/replication_logic.go

```diff
@@ -234,25 +234,38 @@ func NewPlanner(secsPerState *prometheus.HistogramVec, bytesReplicated *promethe
 		promBytesReplicated: bytesReplicated,
 	}
 }
-func resolveConflict(conflict error) (path []*pdu.FilesystemVersion, msg string) {
+func resolveConflict(conflict error, policy PlannerPolicy) (initialSend bool, path []*pdu.FilesystemVersion, msg string) {
 	if noCommonAncestor, ok := conflict.(*ConflictNoCommonAncestor); ok {
 		if len(noCommonAncestor.SortedReceiverVersions) == 0 {
-			// TODO this is hard-coded replication policy: most recent snapshot as source
+			// NOTE: I did not keep this in sync with listStaleFiltering, it isn't clear to me what the right thing to do is.
 			// NOTE: Keep in sync with listStaleFiltering, it depends on this hard-coded assumption
-			var mostRecentSnap *pdu.FilesystemVersion
-			for n := len(noCommonAncestor.SortedSenderVersions) - 1; n >= 0; n-- {
-				if noCommonAncestor.SortedSenderVersions[n].Type == pdu.FilesystemVersion_Snapshot {
-					mostRecentSnap = noCommonAncestor.SortedSenderVersions[n]
-					break
+			if policy.InitiallyAllSnapshots {
+				var allSnapshots []*pdu.FilesystemVersion
+				for n := 0; n < len(noCommonAncestor.SortedSenderVersions); n++ {
+					if noCommonAncestor.SortedSenderVersions[n].Type == pdu.FilesystemVersion_Snapshot {
+						allSnapshots = append(allSnapshots, noCommonAncestor.SortedSenderVersions[n])
+					}
 				}
+				if len(allSnapshots) == 0 {
+					return true, nil, "no snapshots available on sender side"
+				}
+				return true, allSnapshots, "start replication at the oldest snapshot"
+			} else {
```

Comment:
Since you're returning in the if branch, you don't need the else branch.

https://github.com/golang/go/wiki/CodeReviewComments#indent-error-flow (applies to non-error handling code as well)

Reply:
This is policy code, not error handling. I prefer policy code to be a decision tree <=> if + else